core/src/main/java/com/sequenceiq/cloudbreak/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -248,4 +254,123 @@ private boolean isPreferredSubnetsSpecifiedForScaling(NetworkScaleDetails networ
+            List<String> subnetIds = new ArrayList<>();
+            Set<String> availabilityZones = new HashSet<>();
+            for (InstanceMetaData instanceMetaData : instanceGroup.getNotDeletedAndNotZombieInstanceMetaDataSet()) {
+                if (isNullOrEmpty(instanceMetaData.getSubnetId())) {
+                    GroupPlacement groupPlacement = getGroupPlacementForInstanceGroupName(instanceMetaData.getInstanceGroupName(),
+                            instanceMetaData.getInstanceGroup().getHints(), groupPlacementList);
+                    Integer numberOfInstanceInASubnet = searchTheSmallestInstanceCountForUsage(groupPlacement.getSubnetUsage());
+                    String leastUsedSubnetId = searchTheSmallestUsedID(groupPlacement.getSubnetUsage(), numberOfInstanceInASubnet);
+
+                    LOGGER.info("{} instance (Group {}) placed at {}", instanceMetaData.getInstanceGroupName(), groupPlacement.getName(),
+                            groupPlacement.getAvailabilityZone());
+                    instanceMetaData.setSubnetId(leastUsedSubnetId);
+                    instanceMetaData.setAvailabilityZone(groupPlacement.getAvailabilityZone());
+
+                    groupPlacement.addInstanceGroupName(instanceMetaData.getInstanceGroupName());
+                    groupPlacement.increaseSubnetUsage(leastUsedSubnetId);
+                    subnetIds.add(leastUsedSubnetId);
+                    availabilityZones.add(groupPlacement.getAvailabilityZone());
+                }
+            }
+            updateInstanceGroup(instanceGroup, subnetIds, availabilityZones);
+        }
+    }
+
+    private void updateInstanceGroup(InstanceGroup instanceGroup, List<String> subnetIds, Set<String> azNames) {
+        Map<String, Object> networkAttributes = Optional
+                .of(instanceGroup.getInstanceGroupNetwork())
+                .map(InstanceGroupNetwork::getAttributes)
+                .map(Json::getMap)
+                .orElse(new HashMap<>());
+        networkAttributes.put(SUBNET_IDS, subnetIds);
+        instanceGroup.getInstanceGroupNetwork().setAttributes(new Json(networkAttributes));
+
+        Set<AvailabilityZone> availabilityZones = new HashSet<>();
+        azNames.forEach(az -> {
+            AvailabilityZone availabilityZone = new AvailabilityZone();
+            availabilityZone.setAvailabilityZone(az);
+            availabilityZone.setInstanceGroup(instanceGroup);
+            availabilityZones.add(availabilityZone);
+
+        });
+        instanceGroup.setAvailabilityZones(availabilityZones);
+    }
+
+    private GroupPlacement getGroupPlacementForInstanceGroupName(String instanceGroupName, Set<String> hints, Set<GroupPlacement> groupPlacements) {
+        if (hints == null || hints.isEmpty()) {
+            return groupPlacements
+                    .stream()
+                    .findFirst()
+                    .get();
+        }
```

Comment:
if hints is not provided for an instance group in template then all nodes will be put in the same availability zone. It does not seem correct.